File: package/main/src/Error/matchResult.ts
import type { Result } from "@/Error/safeExecute";

/**
 * Pattern-matches on a Result, applying the appropriate handler
 * depending on whether the Result is a success or an error.
 *
 * @param result - The Result to match on.
 * @param handlers - An object with onSuccess and onError callbacks.
 * @param handlers.onSuccess - Called with the value if the Result is a success.
 * @param handlers.onError - Called with the error if the Result is an error.
 * @returns The return value of whichever handler is invoked.
 *
 * @example
 * ```typescript
 * const result = { type: "success", value: 42 } as const;
 * matchResult(result, {
 *   onSuccess: (v) => `Got ${v}`,
 *   onError: (e) => `Failed: ${e}`,
 * });
 * // "Got 42"
 * ```
 */
export const matchResult = <V, E, S, F>(
  result: Result<V, E>,
  handlers: {
    onSuccess: (value: V) => S;
    onError: (error: E) => F;
  },
): S | F => {
  if (result.type === "success") {
    return handlers.onSuccess(result.value);
  }
  return handlers.onError(result.error);
};

File: package/umt_python/src/error/match_result.py
from collections.abc import Callable
from typing import TypeVar

from .safe_execute import Result, Success

V = TypeVar("V")
E = TypeVar("E", bound=Exception)
S = TypeVar("S")
F = TypeVar("F")


def match_result(
    result: Result[V, E],
    on_success: Callable[[V], S],
    on_error: Callable[[E], F],
) -> S | F:
    """
    Pattern-matches on a Result, applying the appropriate handler
    depending on whether the Result is a success or an error.

    Args:
        result: The Result to match on
        on_success: Called with the value if the Result is a success
        on_error: Called with the error if the Result is an error

    Returns:
        The return value of whichever handler is invoked

    Example:
        >>> from .safe_execute import Success
        >>> success = Success(value=42)
        >>> match_result(success, lambda v: f"Got {v}", lambda e: f"Failed: {e}")
        'Got 42'
    """
    if isinstance(result, Success):
        return on_success(result.value)
    return on_error(result.error)
